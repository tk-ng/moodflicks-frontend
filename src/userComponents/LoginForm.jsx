import { useFields } from "../hooks/useFields";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { useNavigate } from "react-router";
import { toast } from "react-toastify";

export const LoginForm = ({ login }) => {
	const navigate = useNavigate();

	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await login(formData);
			resetForm();
			navigate("/");
			toast.success(`Welcome back, ${formData.username}!`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="LoginForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Log In</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
