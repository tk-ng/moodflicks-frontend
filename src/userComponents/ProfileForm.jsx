import { useFields } from "../hooks/useFields";
import { useNavigate, useParams, Navigate } from "react-router";
import { Api } from "../api/api";
import { UserContext } from "./UserContext";
import { useContext, useState, useEffect } from "react";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";
import { toast } from "react-toastify";

export const ProfileForm = () => {
	const { currUser, setCurrentUser } = useContext(UserContext);

	const params = useParams();
	const naviagte = useNavigate();
	const [profileUser, setProfileUser] = useState();

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	const [
		formData,
		handleChange,
		resetForm,
		formErrors,
		setFormErrors,
		setFormData,
	] = useFields({
		firstName: "",
		lastName: "",
		email: "",
	});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			const updatedUser = await Api.saveChanges({
				username: profileUser.username,
				data: formData,
			});
			if (currUser.username === profileUser.username)
				setCurrentUser((currentUser) => ({
					...currentUser,
					data: updatedUser,
				}));
			resetForm();
			toast.success("Profile Updated Successfully");
			naviagte(`/users/${profileUser.username}`);
		} catch (err) {
			setFormErrors(err);
		}
	};

	async function getUserInfo(username) {
		try {
			setIsLoading(true);
			setNotFound(false);

			const userResult = await Api.getUser(username);
			setProfileUser(userResult);
			setFormData({
				firstName: userResult.firstName,
				lastName: userResult.lastName,
				email: userResult.email,
			});
		} catch (err) {
			console.error("User not found:", err);
			setNotFound(true);
		} finally {
			setIsLoading(false);
		}
	}

	useEffect(() => {
		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);

		return <Navigate to="/" />;
	}

	if (!currUser.isAdmin && params.username !== currUser.username) {
		toast.warning(`You are not logged in as ${params.username}.`);
		return <Navigate to={`/users/${currUser.username}`} />;
	}

	return (
		<div className="ProfileForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Profile</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={params.username}
										disabled
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Save Changes</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
