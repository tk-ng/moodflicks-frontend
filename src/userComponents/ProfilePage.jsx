import { useContext, useEffect, useState } from "react";
import { UserContext } from "./UserContext";
import { Navigate, useParams } from "react-router";
import { Api } from "../api/api";
import { Button } from "reactstrap";
import { useNavigate } from "react-router";
import { MoodList } from "../moodComponents/MoodList";
import { toast } from "react-toastify";

export const ProfilePage = () => {
	const params = useParams();
	const navigate = useNavigate();
	const [user, setUser] = useState();
	const [moods, setMoods] = useState([]);
	const { currUser } = useContext(UserContext);

	const [isLoading, setIsLoading] = useState(true);
	const [notFound, setNotFound] = useState(false);

	useEffect(() => {
		async function getUserInfo(username) {
			try {
				setIsLoading(true);
				setNotFound(false);
				const [userResult, moodsResult] = await Promise.all([
					Api.getUser(username),
					Api.getMoodsByUser(username),
				]);
				setUser(userResult);
				setMoods(moodsResult);
			} catch (err) {
				console.error("User not found:", err);
				setNotFound(true);
			} finally {
				setIsLoading(false);
			}
		}

		getUserInfo(params.username);
	}, [params.username]);

	if (isLoading) return <p>Loading...</p>;
	if (notFound) {
		toast.warning(`username: ${params.username} could not be found.`);
		return <Navigate to="/" />;
	}

	return (
		<div className="text-light">
			<span className="display-6">{user.username}</span>
			{currUser && user.username === currUser.username && (
				<Button
					onClick={() => navigate(`/users/${user.username}/edit`)}
					className="m-3 "
				>
					Edit Profile
				</Button>
			)}
			<div>
				{moods.length ? (
					<>
						<h3> Mood-based movie lists, curated by {user.username}</h3>
						<MoodList moods={moods} hideCreatedBy={true} />
					</>
				) : (
					<span>{user.username} has not created any mood lists yet.</span>
				)}
			</div>
		</div>
	);
};
