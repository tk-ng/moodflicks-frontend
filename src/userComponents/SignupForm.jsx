import { useFields } from "../hooks/useFields";
import { useNavigate } from "react-router";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";

export const SignupForm = ({ signup }) => {
	const naviagte = useNavigate();
	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
			firstName: "",
			lastName: "",
			email: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await signup(formData);
			resetForm();
			naviagte("/");
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="SignupForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Sign Up</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
