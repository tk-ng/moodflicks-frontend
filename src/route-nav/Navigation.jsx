import {
	Navbar,
	Nav,
	NavbarBrand,
	NavItem,
	NavbarToggler,
	Collapse,
} from "reactstrap";
import { NavLink, Link } from "react-router";
import { useContext, useState } from "react";
import { UserContext } from "../userComponents/UserContext";

export const Navigation = ({ logout }) => {
	const { currUser } = useContext(UserContext);

	const [isOpen, setIsOpen] = useState(false);
	const toggle = () => setIsOpen(!isOpen);
	return (
		<div className="Navigation mb-3">
			<Navbar expand="md" dark={true}>
				<NavbarBrand href="/">MoodFlicks</NavbarBrand>
				<NavbarToggler onClick={toggle} />
				<Collapse isOpen={isOpen} navbar>
					<Nav navbar className="ms-auto">
						{currUser ? (
							<>
								<NavItem>
									<NavLink className="nav-link" to="/moods/new">
										Create Mood
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink className="nav-link" to="/moods">
										Moods
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink className="nav-link" to="/movies">
										Movies
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink
										className="nav-link"
										to={`/users/${currUser.username}`}
									>
										Profile
									</NavLink>
								</NavItem>
								<NavItem>
									<Link className="nav-link" to="/" onClick={logout}>
										Log out
									</Link>
								</NavItem>
							</>
						) : (
							<>
								<NavItem>
									<NavLink className="nav-link" to="/moods">
										Moods
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink className="nav-link" to="/movies">
										Movies
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink className="nav-link" to="/login">
										Login
									</NavLink>
								</NavItem>
								<NavItem>
									<NavLink className="nav-link" to="/signup">
										Sign Up
									</NavLink>
								</NavItem>
							</>
						)}
					</Nav>
				</Collapse>
			</Navbar>
		</div>
	);
};
