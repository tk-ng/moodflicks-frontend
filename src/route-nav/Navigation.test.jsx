import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { Navigation } from "./Navigation";
import { UserContext } from "../userComponents/UserContext";

describe("Renders the Navigation component", function () {
	const logout = vi.fn();
	it("renders without crashing", function () {
		const { container, asFragment } = render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<Navigation logout={logout} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
		screen.debug();
	});

	it("renders the logged in navbar", function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: { username: "testuser" },
					}}
				>
					<Navigation logout={logout} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(
			screen.getByText("Create Mood", { selector: "a" })
		).toBeInTheDocument();
		expect(screen.getByText("Profile", { selector: "a" })).toBeInTheDocument();
		expect(screen.getByText("Log out", { selector: "a" })).toBeInTheDocument();
		expect(
			screen.queryByText("Sign Up", { selector: "a" })
		).not.toBeInTheDocument();
		expect(
			screen.queryByText("Login", { selector: "a" })
		).not.toBeInTheDocument();
	});

	it("matches snapshot when logged out", function () {
		const { asFragment } = render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: null,
					}}
				>
					<Navigation logout={logout} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders the logged out navbar", function () {
		render(
			<MemoryRouter>
				<UserContext.Provider
					value={{
						currUser: null,
					}}
				>
					<Navigation logout={logout} />
				</UserContext.Provider>
			</MemoryRouter>
		);
		expect(
			screen.queryByText("Create Mood", { selector: "a" })
		).not.toBeInTheDocument();
		expect(
			screen.queryByText("Profile", { selector: "a" })
		).not.toBeInTheDocument();
		expect(
			screen.queryByText("Log out", { selector: "a" })
		).not.toBeInTheDocument();
		expect(screen.getByText("Sign Up", { selector: "a" })).toBeInTheDocument();
		expect(screen.getByText("Login", { selector: "a" })).toBeInTheDocument();
	});
});
