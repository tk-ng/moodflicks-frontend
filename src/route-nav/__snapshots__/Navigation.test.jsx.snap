// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders the Navigation component > matches snapshot when logged out 1`] = `
<DocumentFragment>
  <div
    class="Navigation mb-3"
  >
    <nav
      class="navbar navbar-expand-md navbar-dark"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          MoodFlicks
        </a>
        <button
          aria-label="Toggle navigation"
          class="navbar-toggler"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
        >
          <ul
            class="ms-auto navbar-nav"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/moods"
              >
                Moods
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/movies"
              >
                Movies
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/login"
              >
                Login
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/signup"
              >
                Sign Up
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;

exports[`Renders the Navigation component > renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="Navigation mb-3"
  >
    <nav
      class="navbar navbar-expand-md navbar-dark"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          MoodFlicks
        </a>
        <button
          aria-label="Toggle navigation"
          class="navbar-toggler"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
        >
          <ul
            class="ms-auto navbar-nav"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/moods/new"
              >
                Create Mood
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/moods"
              >
                Moods
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/movies"
              >
                Movies
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/users/testuser"
              >
                Profile
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-discover="true"
                href="/"
              >
                Log out
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
